pkg/users/users.go

```diff
@@ -80,113 +83,121 @@ func DeleteUser(jxClient versioned.Interface, ns string, userName string) error
 	return err
 }
 
-// Resolve does the heavy lifting for user resolution.
-// This function is not normally called directly but by a dedicated user resolver (e.g. GitUserResolver)
-// * checking the user custom resources to see if the user is present there
-// * calling selectUsers to try to identify the user
-// as often user info is not complete in a git response
-func Resolve(id string, providerKey string, jxClient versioned.Interface,
```

Comment:
This is a public API in JX and cannot be changed in this way as it will break calling code. Please keep the API signature as it is and if you need to introduce a different signature add a new function.

The signature of this API is purposefully designed to not be tied to the `GitUser` interface so that other systems that handle users (e.g. Slack) can use it - and the slack bot does indeed use it.

Reply:
I think there are several things wrong with the current code, but to move this forward I created #6366 with the minimal required changes.

I think we will have a very hard time to refactor and improve our code base if we treat every exported function as API. It is unfortunate, that we don't have a clear separation between internal code and actual code we want others (e.g. Apps) to use. 

Regarding the function specifically, I have a very hard time with its design. Both functions parameters and return values are very cryptic and passing the function pointer seem very specific.